### ISSUE TITLE
gossip: sentinel key expiration can lead to a thundering herd of work

### ISSUE THREAD

**Issue by aayushshah15** [cockroachdb/cockroach#76121]
In https://github.com/cockroachlabs/support/issues/1408, we saw a node in the cluster completely saturate its CPU and eventually OOM due to a thundering herd of gossip related work that was triggered by the gossip [sentinel key](https://github.com/cockroachdb/cockroach/blob/8905b9bbb7f0c5a8d623f0fbf5427984a1084056/pkg/gossip/keys.go#L54-L58) hitting its expiration.  

Generally, expiration of the sentinel key means that the node is likely partitioned away from the cluster and should try to establish other gossip peers so that it can keep up to date with the disposition of meta1.
```
W220128 16:18:57.743724 429 2@gossip/gossip.go:1491 ⋮ [n228] 66592 first range unavailable; trying remaining resolvers
```
However, the gossip server checks the validity of the sentinel every [2 seconds](https://github.com/cockroachdb/cockroach/blob/2a3fe98da5ad3869ec3d081777cb8fc875b7c360/pkg/gossip/gossip.go#L96-L99), and if it finds that the sentinel key is expired, it [establishes a new gossip client to another node in the cluster](https://github.com/cockroachdb/cockroach/blob/2a3fe98da5ad3869ec3d081777cb8fc875b7c360/pkg/gossip/gossip.go#L1310-L1313). This means that a node that hasn't learned about meta1 in the last ~4 seconds is going to potentially continuously establish new gossip clients. We saw logs of the following form confirming this: 
```
gossip client (5/4 cur/max conns)
...
gossip client (9/4 cur/max conns)
...
gossip client (13/4 cur/max conns)
...
gossip client (20/4 cur/max conns)
```

These gossip clients can potentially be established at an interval of 2 seconds, but are only culled (one at a time) [every minute](https://github.com/cockroachdb/cockroach/blob/2a3fe98da5ad3869ec3d081777cb8fc875b7c360/pkg/gossip/gossip.go#L106-L109). 

@nvanbenschoten noticed that there was a sharp inflection in the rate of new gossip info objects being received by this errant node, which lines up exactly with the first gossip message from above indicating the expiration of the sentinel key. 
![gossip infos received n228](https://user-images.githubusercontent.com/10788754/152675007-9d7158fc-1f73-48b1-964b-ee49e22b7c32.png)

In a large cluster, like the one the escalation was about, this proved to be very hazardous and pushed the node into what seems like a "metastable" regime where it was constantly behind on processing gossip info objects, leading to repeated expiration of its sentinel key's TTL, leading more gossip clients being established, leading to more gossip infos being received.
![image](https://user-images.githubusercontent.com/10788754/152674713-a5d03a02-731f-4c01-a222-d5523665cc2b.png)

cc @cockroachdb/kv-notifications

**Comment by tbg** [comment]
Nice sleuthing! To reproduce this without actually overloading the system, perhaps we can insert a `Sleep()` in the right place.

**Comment by petermattis** [comment]
Ditto on the nice sleuthing! Are the memory and CPU usage increases directly tied to something in gossip-land? I'd be a bit surprised by that. I don't recall gossip connections being that heavyweight.

**Comment by knz** [comment]
Do you think this could also explain #51838 ?

**Comment by tbg** [comment]
@knz interesting connection, especially [this graph](https://user-images.githubusercontent.com/5198575/116718909-7952c700-a9a8-11eb-90e7-be268eb60ab5.png) has a lot of CPU in the infostore.

**Comment by aayushshah15** [comment]
> Are the memory and CPU usage increases directly tied to something in gossip-land?

@petermattis: Yeah, the heap profiles leading up to the OOM show more and more memory allocations coming from [here](https://github.com/cockroachdb/cockroach/blob/229d47ba478d0b511c6d49cb0b8952f94188b90e/pkg/gossip/infostore.go#L329). Quoting myself from a comment on the support issue:
>It seems to me that the memory pressure buildup (and the eventual OOM) is an interesting consequence of the CPU saturation. The processCallbacks function (and its callee runCallbacks) seem to schedule their work into this [slice here](https://github.com/cockroachdb/cockroach/blob/229d47ba478d0b511c6d49cb0b8952f94188b90e/pkg/gossip/infostore.go#L342-L344). All that memory buildup seems to be from this slice growing unboundedly.

>The memory allocated to this seems like its only reclaimed when the infostore goroutine actually [processes the work](https://github.com/cockroachdb/cockroach/blob/229d47ba478d0b511c6d49cb0b8952f94188b90e/pkg/gossip/infostore.go#L179-L180). This means that if, for whatever reason (like CPU saturation), the work is not getting processed, we will see buildup of memory pressure.

The CPU profile from a node that failed the same way has a lot of time spent in replica rebalancing (since they were adding new nodes to the cluster) and a significant chunk in gossip-land.

**Comment by petermattis** [comment]
Ah, that makes sense now with regard to the memory buildup, @aayushshah15. The gossip code is some of the oldest code in CRDB and I think we'd structure it differently nowadays. Fundamentally, it is bad that we're queueing a callback to be invoked every time a key matching the callback's matcher is received. Instead, we should only queue the callback if it is not already queued. The locking for that might be tricky, though. Separately, we need to fix gossip to not add new connections so quickly when the sentinel key isn't found (or to more aggressively cull old connections when doing so).

**Comment by ambareesh-jayakumari** [comment]
@aayushshah15 / @knz  Is there a way of preventing this from a happening in a cluster where it is happening fairly consistently? I've a large cluster where intermittently some nodes (different one each time), keep OOM-ing out and symptoms are very similar to what's discussed here. 
I'm looking for maybe config-changes (or any other tunables) to control it somehow

**Comment by a-robinson** [comment]
I ran into a version of this recently as well. I've thrown up a test that reproduces the fairly convoluted worst-case scenario along with some ideas for mitigating it in #121361 if anyone's interested.
